clients/pkg/promtail/targets/gcplog/push_translation.go

```diff
@@ -37,29 +28,39 @@ type PushMessage struct {
 	Subscription string `json:"subscription"`
 }
 
-// translate converts a GCP PushMessage into a loki api.Entry. It is responsible for decoding the log line (contained in the Message.Data)
-// attribute, using the incoming timestamp if necessary, and formatting and passing down the incoming Message.Attributes
-// if relabel is configured.
+func (pm PushMessage) Validate() error {
+	if pm.Message.Data == "" {
+		return fmt.Errorf("push message has no data")
+	}
+	if pm.Message.ID == "" {
+		return fmt.Errorf("push message has no ID")
+	}
+	if pm.Subscription == "" {
+		return fmt.Errorf("push message has no subscription")
+	}
+	return nil
+}
+
+// translate converts a GCP PushMessage into a loki api.Entry. It parses the push-specific labels, and delegates the rest to parseGCPLogsEntry.
 func translate(m PushMessage, other model.LabelSet, useIncomingTimestamp bool, relabelConfigs []*relabel.Config, xScopeOrgID string) (api.Entry, error) {
-	// mandatory label for gcplog
+	// Collect all push-specific labels. Every one of them is first configured as optional, and the user
+	// can relabel it if needed.
 	lbs := labels.NewBuilder(nil)
 	lbs.Set("__gcp_message_id", m.Message.ID)
-
-	// labels from gcp log entry. Add it as internal labels
+	lbs.Set("__gcp_subscription_name", m.Subscription)
 	for k, v := range m.Message.Attributes {
 		lbs.Set(fmt.Sprintf("__gcp_attributes_%s", convertToLokiCompatibleLabel(k)), v)
 	}
 
+	// Apply relabeling
```

Comment:
Do you think there would be any harm in removing the `Apply relabeling` and `Drop internal or invalid labels` logic to let the `parseGCPLogsEntry` call further down take care of it? At a glance it looks like the logic is the same.

Reply:
Because of how `parseGCPLogsEntry` uses the push-specific labels we pass down, they wouldn't be affected by the relabeling + drop internal it does. [This](https://github.com/thepalbi/loki/blob/1da24900d668fa7afa33e4f00e8c92d9f2e4078c/clients/pkg/promtail/targets/gcplog/formatter.go#L87) is the point they are used, which is after relabel + drop internal.
We could could instead pass them like:
```
func parseGCPLogsEntry(
	data []byte,
	other model.LabelSet,
	useIncomingTimestamp bool,
	relabelConfig []*relabel.Config,
        additionalInternalLabels ...model.LabelSet
) (api.Entry, error)
```
That way we differentiate them from the ones that must be merged after like the fixed labels, or the `__tenant_id__`. wdyt? It would just be to avoid repeating code, since the sum of iterations between both loops is the same

Reply:
If it's not too much of a work, I would also prefer to remove the repeating code from `translate` and let `parseGCPLogsEntry` handles it. 

Or we can move relabel and merge to another helper and just call it in two places? will that work?